cluster-autoscaler/proposals/plugable-provider-grpc.md

```diff
@@ -0,0 +1,604 @@
+  map<string, k8s.io.apimachinery.pkg.api.resource.Quantity extraResources = 5;
+}
+
+message NewNodeGroupResponse {
+  // NodeGroup created by the cloud provider.
+ NodeGroup nodeGroup = 1;
+}
+```
+
+### GetResourceLimiter
+
+GetResourceLimiter holds a struct containing limits (max, min) for resources (cores, memory etc.).
+
+```protobuf
+message GetResourceLimiterRequest {
+  // Intentionally empty.
+}
+
+message GetResourceLimiterResponse {
+ // Contains the minimum limits for resources (cores, memory etc.).
+ map<string, int64> minLimits = 1;
+
+ // Contains the maximum limits for resources (cores, memory etc.).
+ map<string, int64> maxLimits = 2;
+}
+```
+
+### Refresh
+
+Refresh is called before every main loop and can be used to dynamically update cloud provider state.
+In particular the list of node groups returned by NodeGroups can change as a result of this action.
+
+```protobuf
+message RefreshRequest {
+  // Intentionally empty.
+}
+
+message RefreshResponse {
+// Intentionally empty.
+}
+```
+
+### Cleanup
+
+Cleanup cleans up open resources before the cloud provider is destroyed, i.e. go routines etc.
+
+```protobuf
+message CleanupRequest {
+  // Intentionally empty.
+}
+
+message CleanupResponse {
+// Intentionally empty.
+}
+```
+
+### Name
+
+Name stores the name of the cloud provider.
+
+```protobuf
+message NameRequest {
+  // Intentionally empty.
+}
+
+message NameResponse {
+  // Name of the node group
+ string name = 1;
+}
+```
+
+### NodeGroup
+
+```protobuf
+message NodeGroup {
+  // Name of the node group on the cloud provider
+  string name = 1;
+
+  // ID of the node group on the cloud provider
+  string id = 1;
+
+  // MinSize of the node group on the cloud provider
+  int32 minSize = 2;
+
+  // MaxSize of the node group on the cloud provider
+  int32 maxSize = 3;
+
+  // Exist reports if the node group really exists on the cloud provider
+  bool exist = 4;
+
+  // Autoprovisioned returns true if the node group is autoprovisioned
+  bool autoProvisioned = 5;
+
+  // Debug returns a string containing all information regarding this node group.
+  string debug = 6;
+}
+```
+
+### NodeGroupTargetSize
+
+NodeGroupTargetSize returns the current target size of the node group.
+
+```protobuf
+message NodeGroupTargetSizeRequest {
+  // ID of the group node on the cloud provider
+  string id = 1;
+}
+
+message NodeGroupTargetSizeResponse {
+  // TargetSize
+ int32 targetSize = 1;
+}
+```
+
+### NodeGroupIncreaseSize
+
+NodeGroupIncreaseSize describes the message type of an action that increases the size of the node group.
+
+```protobuf
+message NodeGroupIncreaseSizeRequest {
+ int32 delta = 1;
+
+ // ID of the group node on the cloud provider
+ string id = 2;
+}
+
+message NodeGroupIncreaseSizeResponse {
+  // Intentionally empty.
+}
+```
+
+### NodeGroupDeleteNodes
+
+NodeGroupDeleteNodes deletes nodes from this node group.
+
+```protobuf
+message NodeGroupDeleteNodesRequest {
+ repeated k8s.io.api.core.v1.Node nodes = 1;
```

Comment:
Again, I don't think there is any need to send the whole node object (which may be outdated, etc). I think we should consider defining some sort of node identifier and use this.

Reply:
It makes sense. Good call!